### PR TITLE
netsurf: create alias NodeExternal

### LINKED ISSUE #109: netsurf: create an internal alias `NodeExternal` to avoid using c pointer
> We already talked about creating an internal alias `NodeExternal` to avoid using c pointer. Let's do that in a separate PR.

> I think once we have the NodeExternal type we can remove those comments who are not so useful.

_Originally posted by @francisbouvier in https://github.com/Browsercore/browsercore/pull/100#discussion_r1414114553_

### PR DESCRIPTION
**Pull request by krichprollsch**
fix #109

### CODE REVIEW

File: src/netsurf.zig [hunk 1]
```diff
@@ -479,16 +466,26 @@ pub fn nodeListLength(nodeList: *NodeList) !u32 {
...
     return @as(*Node, @ptrCast(n));
 }
 
+// NodeExternal is the libdom public representation of a Node.
+// Since we use the internal representation (dom_node_internal), we declare
+// here a private version useful for some netsurf function call.
+const NodeExternal = [*c]c.dom_node;
+
+// Convert a parser pointer to a NodeExternal pointer.
+fn toNodeExternal(comptime T: type, v: *T) NodeExternal {
```
Comment (francisbouvier) [inline src/netsurf.zig]: Here you could use `v: anytype` but I'm not sure what is better. anytype is shorter but your version in more explicit.

File: src/netsurf.zig [hunk 2]
```diff
@@ -479,16 +466,26 @@ pub fn nodeListLength(nodeList: *NodeList) !u32 {
...
     if (n == null) return null;
 
-    // cast [*c]c.dom_node into *Node
     return @as(*Node, @ptrCast(n));
 }
 
+// NodeExternal is the libdom public representation of a Node.
+// Since we use the internal representation (dom_node_internal), we declare
+// here a private version useful for some netsurf function call.
+const NodeExternal = [*c]c.dom_node;
```
Comment (francisbouvier) [inline src/netsurf.zig]: I would have put this before declaring `Node`, so we do not use types that haven't been yet declared.

File: src/netsurf.zig [hunk 3]
```diff
@@ -347,24 +347,22 @@ pub fn namedNodeMapGetLength(nnm: *NamedNodeMap) !u32 {
 }
 
 pub fn namedNodeMapItem(nnm: *NamedNodeMap, index: u32) !?*Attribute {
-    var n: [*c]c.dom_node = undefined;
+    var n: NodeExternal = undefined;
     const err = c._dom_namednodemap_item(nnm, index, &n);
     try DOMErr(err);
```
Comment (francisbouvier) [inline src/netsurf.zig]: Do we need those line returns now?

File: src/netsurf.zig [hunk 4]
```diff
@@ -347,24 +347,22 @@ pub fn namedNodeMapGetLength(nnm: *NamedNodeMap) !u32 {
...
 
     if (n == null) return null;
```
Comment (francisbouvier) [inline src/netsurf.zig:400]: Same here, do we need those line returns now?

File: src/netsurf.zig (hunk 1 above)
Comment (krichprollsch) [reply src/netsurf.zig]: IMO it's not the same: using `v: *T` ensures `v` is a pointer to a value of type `T`.

If I use `v: anytype` I would be allowed to pass anything as `v` no?

File: src/netsurf.zig (hunk 2 above)
Comment (krichprollsch) [reply src/netsurf.zig]: I'm not sure to understand, `pub const Node = c.dom_node_internal;` is declared after, see https://github.com/Browsercore/browsercore/pull/112/files#diff-eff26b7fb15b686657ba5f430ad7ba96d4a1842a6d51efc0e804f9033ffc66f9L493

File: src/netsurf.zig (hunk 1 above)
Comment (francisbouvier) [reply src/netsurf.zig]: Well in anycase not passing a ptr will result a compile error, as `@alignCast` requires a ptr. At the end it's the same, your version is more explicit, the anytype version is more concise, I'm OK with boh.

File: src/netsurf.zig (hunk 2 above)
Comment (francisbouvier) [reply src/netsurf.zig]: Yes sorry. It's more the namednodemap that could be after but let's keep it like that.

Comment (krichprollsch) [reply src/netsurf.zig]: moved

**Pull request merged**